### ISSUE TITLE
GRC: Estudo padrão de codificação NOAA

### ISSUE THREAD

**Issue by joaomas** [monanadmin/monan#280]
### User Story

COMO Saulo,
QUEREMOS Fazer um estudo do padrão de codificação usado na NOAA,
PARA Alinhar os padrões utilizado no MONAN.

### Critérios de aceite de conclusão da Issue

- Fazer estudo do padrão de codificação da NOAA.
- Este estudo será utilizado na issue de comparação entre os padrões #265 

### Detalhamento adicional da atividade

_-

### Impactos

_-

### Dependências

_-

### Solução

**Coding standards - Coding Standards section of the NOAA** [1]

STATUS: The proposed standards here need to be reconciled with Environmental Equivalence 2. The section should be updated to reflect the standards that NCEP/EMC intends to follow.

| **ID** | **Type**    | **Item**                                                                                                                                                                                                                                                                                                                                              | **Reason**                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 | **Source**               | **Status** |
| ------ | ----------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------------------------ | ---------- |
| CS1    | Required    | All modules or subroutines will contain “implicit none”                                                                                                                                                                                                                                                                                               | Assists in writing bug-free code.Understanding implicit type rules is difficult and arcane.Understanding where a variable comes from (local, input argument list, module) is more difficult with implicit typing                                                                                                                                                                                                                                                                                                                                                                                           | GMTB, GSM                |            |
| CS2    | Required    | All arguments to subprograms will contain the “intent” attribute                                                                                                                                                                                                                                                                                      | Assists readers in understanding whether a variable is read-only (intent(in)), read/write (intent(inout)), or effectively uninitialized (intent(out)).A compiler error will result if code attempts to use a variable differently than specified in its “intent”.Declared variables without the “intent” attribute can be understood to be local.                                                                                                                                                                                                                                                          | GMTB, NUOPC PI Team, GSM |            |
| CS3    | Required    | No modules or subroutines will violate the Fortran 2008 standard                                                                                                                                                                                                                                                                                      | Makes porting to a new compiler easier to near trivial.Example: gfortran by default enforces the standard that free-form source lines will not exceed 132 characters. Some compilers by default allow line lengths to exceed this value. Attempts to port codes with line lengths greater than 132 may encounter difficulty.                                                                                                                                                                                                                                                                               | GMTB                     |            |
| CS4    | Required    | All local and argument list variables will have a comment explaining the meaning of the variable. An in-line comment on the declaration line is sufficient                                                                                                                                                                                            | Allows readers unfamiliar with the code to more quickly understand how the code works.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     | GMTB, NUOPC PI Team, GSM |            |
| CS5    | Required    | All modules and subprograms will have a documentation block describing functionality                                                                                                                                                                                                                                                                  | Promotes understanding of algorithms and code structure by new users                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       | GMTB, NUOPC PI Team, GSM |            |
| CS6    | Required    | Common blocks are disallowed                                                                                                                                                                                                                                                                                                                          | Deprecated Fortran feature. Modules provide all the functionality of common blocks plus much more.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         | GMTB, NUOPC PI Team, GSM |            |
| CS7    | Required    | A package must be compilable with the gfortran compiler (or gcc for packages coded in C). Runnability and validation can be provided using whatever compiler(s) the developer prefers.                                                                                                                                                                | gfortran (and gcc) is free and ubiquitous, and therefore is an ideal choice for canonical compiler.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        | GMTB                     |            |
| CS8    | Required    | All Fortran source will be free-form                                                                                                                                                                                                                                                                                                                  | Fixed-form source is hard to read and archaic.A 72-column requirement only makes sense for punch cards.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    | GMTB                     |            |
| CS9    | Required    | All public subprograms will be Fortran-callable                                                                                                                                                                                                                                                                                                       | Fortran is the most commonly used language for geophysical models.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         | GMTB                     |            |
| CS10   | Required    | All parameterizations must be thread-safe (except for initialization and finalization methods)                                                                                                                                                                                                                                                        | Many geophysical numerical models are threaded these days, and need to be able to invoke physical parameterizations simultaneously from multiple threads.Example code which is NOT thread-safe: Declare a variable “first” and initialize it to .true. Then test its value and set some static variables if it is .true. This will likely result in wrong answers when run in threaded mode.Solution: Provide an initialization routine which sets the static variables outside of threaded regions.Wikipedia provides a brief overview of thread-safety: https&#x3A;//en.wikipedia.org/wiki/Thread_safety | GMTB, NUOPC PI Team, GSM |            |
| CS11   | Required    | No parameterization will contain a “stop” or “abort” clause                                                                                                                                                                                                                                                                                           | If an error condition arises, it is better to set a flag and let the caller decide how to handle the condition.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            | GMTB, NUOPC PI Team, GSM |            |
| CS12   | Required    | Use of uninitialized variables is disallowed                                                                                                                                                                                                                                                                                                          | Readability.Not all compilers can be made to initialize static or stack variables to a known value (e.g. zero).                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            | GMTB, GSM                |            |
| CS13   | Required    | All array indices must fall within their declared bounds.                                                                                                                                                                                                                                                                                             | Debuggers will fail when “tricks” are employed which reference arrays outside of their declared bounds.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    | GMTB, NUOPC PI Team, GSM |            |
| CS14   | Required    | Multiple runs of the same compiled parameterization given identical input must produce identical output. In the case where randomness is part of the parameterization, a method must be provided to invoke the same random sequence for test reproducibility.                                                                                         | Prevents inadvertent errors.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               | GMTB, GSM                |            |
| CS15   | Required    | The use of compiler flags specifying default precision is disallowed. For example, if 64-bit precision is required, use the “kind=” attribute to specify the precision rather than a compiler flag such as “-r8”                                                                                                                                      | The behavior of flags is compiler-specific, e.g. if the user specifies real\*4 does the -r8 compiler flag mean the variable is real\*4 or real\*8?                                                                                                                                                                                                                                                                                                                                                                                                                                                         | GMTB                     |            |
| CS16   | Recommended | With the exception of common libraries which use a well-defined naming standard for variables and subroutines, all module “use” statements must explicitly state which public entities will be referenced. The MPI library is an example of an acceptable exception: All MPI routines start with “MPI”, so a blank “use mpi” statement is acceptable. | Assists in understanding where various variables and/or functions or subroutines are defined.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              | GMTB                     |            |
| CS17   | Recommended | All code intended for debugging purposes only should be removed prior to submission for inclusion                                                                                                                                                                                                                                                     | Readability                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                | GMTB, GSM                |            |
| CS18   | Required    | All arrays explicitly allocated, must be deallocated when no longer needed                                                                                                                                                                                                                                                                            | Readability. Minimize memory usage.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        | GMTB, GSM                |            |
| CS19   | Recommended | The default visibility rule for module variables and procedures should be “private” (specified by a single “private” statement near the beginning of the module). The “public” attribute is applied to only those entities which are needed by other subprograms or modules.                                                                          | Limiting variable and subprogram scope is good programming practice                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        | GMTB                     |            |
| CS20   | Recommended | Consistent use of case is preferred for Fortran code (text strings excepted).                                                                                                                                                                                                                                                                         | While Fortran is a case-insensitive language, variable “aBc” should also be expressed that way, and not “aBc” in one place, “abc” in another, and “ABC” in another.                                                                                                                                                                                                                                                                                                                                                                                                                                        | GMTB                     |            |
| CS21   | Recommended | A parameterization should contain “init”, “run”, and “finalize” methods. The “run” method must be thread-safe.                                                                                                                                                                                                                                        | Promotes separation of activities which must be done only once at startup or shutdown, from those which are done on multiple time steps.                                                                                                                                                                                                                                                                                                                                                                                                                                                                   | GMTB                     |            |
| CS22   | Recommended | Parameterizations should be able to be invoked in “chunks”, where the calculations are independent of the fastest-varying subscript.                                                                                                                                                                                                                  | Computational performance is the main reason for this preference. Many physical parameterizations in geophysical models contain a dependence in the vertical, which means this dimension is unavailable for vectorization. Vectorization can provide up to a 16X speedup on modern processors.Example: Outer loop over vertical index “k” can contain vertical dependence, but if there is also an inner loop over horizontal index “i” that can be vectorized, the code is likely to run much more efficiently.                                                                                           | GMTB                     |            |
| CS23   | Recommended | The use of “goto” is strongly discouraged, except where no better option is available.                                                                                                                                                                                                                                                                | Modern languages provide better mechanisms to accomplish the same goal in most cases..“goto” promotes “spaghetti” code, which can be unreadable.                                                                                                                                                                                                                                                                                                                                                                                                                                                           | GMTB                     |            |
| CS24   | Recommended | Code and declarations within subprograms, loops, and conditional tests should be indented. Indenting by 2 or 3 or 4 columns is reasonable                                                                                                                                                                                                             | Readability. Particularly important for multiply nested loops and/or “if” tests.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           | GMTB                     |            |
| CS25   | Recommended | Test operators &lt;, &lt;=, >, >=, ==, /= are preferred vs. their deprecated counterparts .lt., .le., .gt., .ge., .eq., .ne.                                                                                                                                                                                                                          | The modern constructs are easier to read, and more understandable for those unfamiliar with legacy code.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   | GMTB, GSM                |            |
| CS26   | Recommended | The use of bare constants (e.g. 2.7) inside of computational regions is strongly discouraged. Instead, a named constant (e.g. some_variable = 2.7) should be declared at the top of the routine or module, along with an in-line comment stating its purpose                                                                                          | Bare constants buried in code is one of the biggest contributors to lack of readability and understanding of how code works. “What the heck does 2.7 mean???” In addition, using a named constant makes it easier to specify precision, e.g. real\*8 some_var = 35.                                                                                                                                                                                                                                                                                                                                        | GMTB                     |            |


**Referências**
[1] https://docs.google.com/document/d/1bjnyJpJ7T3XeW3zCnhRLTL5a3m4_3XIAUeThUPWD9Tg/edit#heading=h.21je8qtnh1hj

### Problemas encontrados

_-

### Conclusão

Regras bem abrangentes, este estudo deverá ser utilizado na issue #265 para a conclusão do estudo comparativo.

**Comment by joaomas** [comment]
@klclaudio migrado as regras do doc diretamente para esta issue.
